File: scripts/generate_data.py
"""Generate data/ CSVs for a ticker using Alpha Vantage (fundamentals) and yfinance (prices)."""

from __future__ import annotations

import argparse
import os
import time
import json
from pathlib import Path
from typing import Optional
from urllib.request import urlopen
from urllib.error import HTTPError, URLError

import pandas as pd

try:
    import yfinance as yf
except Exception as exc:
    raise RuntimeError("yfinance is required: pip install yfinance") from exc

ALPHA_VANTAGE_BASE_URL = "https://www.alphavantage.co/query"


def _ensure_dirs(base_dir: Path):
    fundamentals_dir = base_dir / "fundamentals"
    price_dir = base_dir / "price_data"
    for d in (fundamentals_dir, price_dir):
        d.mkdir(parents=True, exist_ok=True)
    return fundamentals_dir, price_dir


def _quarter_label(dt: pd.Timestamp) -> str:
    quarter = ((dt.month - 1) // 3) + 1
    return f"Q{quarter}"


def _quarter_frame(series: pd.Series, value_name: str) -> pd.DataFrame:
    df = pd.DataFrame({value_name: series.values}, index=series.index)
    df["year"] = df.index.year
    df["quarter"] = df.index.map(_quarter_label)
    return df[["year", "quarter", value_name]]


def _ttm_sum(series: pd.Series) -> pd.Series:
    return series.sort_index().rolling(4).sum()


def _ttm_mean(series: pd.Series) -> pd.Series:
    return series.sort_index().rolling(4).mean()


def _fetch_alpha_vantage(function: str, symbol: str, api_key: str) -> dict:
    url = f"{ALPHA_VANTAGE_BASE_URL}?function={function}&symbol={symbol}&apikey={api_key}"
    try:
        with urlopen(url, timeout=30) as response:
            data = json.loads(response.read().decode("utf-8"))
    except (HTTPError, URLError) as e:
        raise RuntimeError(f"Alpha Vantage API request failed: {e}") from e

    if "Error Message" in data:
        raise RuntimeError(f"Alpha Vantage API error: {data['Error Message']}")
    if "Note" in data:
        raise RuntimeError(f"Alpha Vantage rate limit: {data['Note']}")
    if "Information" in data:
        raise RuntimeError(f"Alpha Vantage: {data['Information']}")
    return data


def _parse_quarterly_reports(data: dict, report_key: str = "quarterlyReports") -> pd.DataFrame:
    reports = data.get(report_key, [])
    if not reports:
        return pd.DataFrame()
    df = pd.DataFrame(reports)
    if "fiscalDateEnding" not in df.columns:
        return pd.DataFrame()
    df["fiscalDateEnding"] = pd.to_datetime(df["fiscalDateEnding"])
    df = df.set_index("fiscalDateEnding").sort_index()
    for col in df.columns:
        if col != "reportedCurrency":
            df[col] = pd.to_numeric(df[col], errors="coerce")
    return df


def _parse_annual_reports(data: dict) -> pd.DataFrame:
    reports = data.get("annualReports", [])
    if not reports:
        return pd.DataFrame()
    df = pd.DataFrame(reports)
    if "fiscalDateEnding" not in df.columns:
        return pd.DataFrame()
    df["fiscalDateEnding"] = pd.to_datetime(df["fiscalDateEnding"])
    df = df.set_index("fiscalDateEnding").sort_index()
    for col in df.columns:
        if col != "reportedCurrency":
            df[col] = pd.to_numeric(df[col], errors="coerce")
    return df


def _pick_series(df: pd.DataFrame, names: list[str]) -> Optional[pd.Series]:
    if df.empty:
        return None
    for name in names:
        if name in df.columns:
            series = df[name].astype("float64")
            if not series.isna().all():
                return series
    return None


def generate_data(ticker: str, data_dir: Path, api_key: Optional[str] = None):
    if api_key is None:
        api_key = os.environ.get("ALPHA_VANTAGE_API_KEY")
    if not api_key:
        raise RuntimeError("Alpha Vantage API key required.")

    base_dir = data_dir / ticker.upper()
    fundamentals_dir, price_dir = _ensure_dirs(base_dir)

    # Prices from yfinance - fetch maximum available history
    print(f"Fetching price data for {ticker}...")
    price = yf.download(ticker, period="max",
                        auto_adjust=False, progress=False, group_by="column")
    if price.empty:
        raise RuntimeError(f"No price data for {ticker}")
    if isinstance(price.columns, pd.MultiIndex):
        price.columns = [c[0] if c[0] else c[1] for c in price.columns]
    price = price.reset_index()
    price.columns = [str(c).lower().replace(" ", "_") for c in price.columns]
    price = price.rename(columns={"adj_close": "adjusted_close"})
    if "date" not in price.columns and "datetime" in price.columns:
        price = price.rename(columns={"datetime": "date"})
    if "close" not in price.columns:
        close_candidates = [c for c in price.columns if "close" in c and c != "adjusted_close"]
        if close_candidates:
            price = price.rename(columns={close_candidates[0]: "close"})
    price["date"] = pd.to_datetime(price["date"]).dt.strftime("%Y-%m-%d")
    price[["date", "close"]].to_csv(price_dir / "daily_prices.csv", index=False)
    print(f"  Saved {len(price)} days of price data")

    # Market data metrics
    price_numeric = price.copy()
    price_numeric["date"] = pd.to_datetime(price_numeric["date"])
    price_numeric = price_numeric.sort_values("date")
    current_price = float(price_numeric.iloc[-1]["close"])
    current_date = price_numeric.iloc[-1]["date"].strftime("%Y-%m-%d")
    last_year_prices = price_numeric.tail(252)
    week_52_high = float(last_year_prices["close"].max())
    week_52_low = float(last_year_prices["close"].min())
    pct_from_high = ((current_price - week_52_high) / week_52_high) * 100

    market_rows = [
        {"metric": "current_price", "value": current_price, "date": current_date},
        {"metric": "52_week_high", "value": week_52_high, "date": current_date},
        {"metric": "52_week_low", "value": week_52_low, "date": current_date},
        {"metric": "pct_from_52week_high", "value": round(pct_from_high, 2), "date": current_date},
    ]
    if len(price_numeric) >= 252:
        price_1y_ago = float(price_numeric.iloc[-252]["close"])
        pct_1y = ((current_price - price_1y_ago) / price_1y_ago) * 100
        market_rows.append({"metric": "pct_change_1y", "value": round(pct_1y, 2), "date": current_date})
    pd.DataFrame(market_rows).to_csv(price_dir / "market_data.csv", index=False)
    print(f"  Saved market data")

    # Fundamentals from Alpha Vantage - save ALL available data
    print(f"Fetching fundamentals from Alpha Vantage...")
    print("  Fetching income statement...")
    income_raw = _fetch_alpha_vantage("INCOME_STATEMENT", ticker, api_key)
    income_q = _parse_quarterly_reports(income_raw)
    income_a = _parse_annual_reports(income_raw)
    time.sleep(12)
    print("  Fetching balance sheet...")
    balance_raw = _fetch_alpha_vantage("BALANCE_SHEET", ticker, api_key)
    balance_q = _parse_quarterly_reports(balance_raw)
    balance_a = _parse_annual_reports(balance_raw)
    time.sleep(12)
    print("  Fetching cash flow statement...")
    cash_raw = _fetch_alpha_vantage("CASH_FLOW", ticker, api_key)
    cash_q = _parse_quarterly_reports(cash_raw)
    cash_a = _parse_annual_reports(cash_raw)

    # Merge quarterly + annual data to maximize history
    # Annual reports go further back; convert to quarterly-like by placing at fiscal date
    def _merge_q_a(q_series, a_series):
        """Extend quarterly series with annual data for dates not covered by quarterly."""
        if q_series is None and a_series is None:
            return None
        if q_series is None:
            return a_series
        if a_series is None:
            return q_series
        # Only keep annual entries that predate the earliest quarterly entry
        earliest_q = q_series.index.min()
        a_before = a_series[a_series.index < earliest_q]
        if a_before.empty:
            return q_series
        return pd.concat([a_before, q_series]).sort_index()

    revenue_q = _pick_series(income_q, ["totalRevenue"])
    revenue_a = _pick_series(income_a, ["totalRevenue"])
    operating_income_q = _pick_series(income_q, ["operatingIncome"])
    operating_income_a = _pick_series(income_a, ["operatingIncome"])
    net_income_q = _pick_series(income_q, ["netIncome", "netIncomeFromContinuingOperations"])
    shares_out_q = _pick_series(balance_q, ["commonStockSharesOutstanding", "commonStock"])
    shares_out_a = _pick_series(balance_a, ["commonStockSharesOutstanding", "commonStock"])
    eps_q = _pick_series(income_q, ["dilutedEPS", "reportedEPS"])
    eps_a = _pick_series(income_a, ["dilutedEPS", "reportedEPS"])

    if eps_q is None:
        diluted_shares = _pick_series(income_q, ["dilutedAverageShares", "dilutedAverageSharesOutstanding"])
        if net_income_q is not None and diluted_shares is not None:
            eps_q = net_income_q / diluted_shares
        elif net_income_q is not None and shares_out_q is not None:
            eps_q = net_income_q / shares_out_q

    fcf_q = _pick_series(cash_q, ["freeCashFlow"])
    if fcf_q is None:
        op_cf = _pick_series(cash_q, ["operatingCashflow"])
        capex = _pick_series(cash_q, ["capitalExpenditures"])
        if op_cf is not None and capex is not None:
            fcf_q = op_cf - capex

    fcf_a = _pick_series(cash_a, ["freeCashFlow"])
    if fcf_a is None:
        op_cf_a = _pick_series(cash_a, ["operatingCashflow"])
        capex_a = _pick_series(cash_a, ["capitalExpenditures"])
        if op_cf_a is not None and capex_a is not None:
            fcf_a = op_cf_a - capex_a

    # For quarterly TTM data, use all available quarterly reports (no tail limit)
    revenue = revenue_q
    if revenue is not None:
        ttm = _ttm_sum(revenue).dropna()
        if not ttm.empty:
            _quarter_frame(ttm, "revenue").to_csv(fundamentals_dir / "revenue.csv", index=False)
            print(f"  Saved {len(ttm)} quarters of revenue")

    operating_income = operating_income_q
    if operating_income is not None and revenue is not None:
        ttm_op = _ttm_sum(operating_income)
        ttm_rev = _ttm_sum(revenue)
        margin = (ttm_op / ttm_rev).dropna()
        if not margin.empty:
            _quarter_frame(margin, "operating_margin").to_csv(fundamentals_dir / "operating_margins.csv", index=False)
            print(f"  Saved {len(margin)} quarters of margins")

    shares_out = shares_out_q
    if shares_out is not None:
        shares_ttm = _ttm_mean(shares_out).dropna()
        if not shares_ttm.empty:
            _quarter_frame(shares_ttm, "shares_outstanding").to_csv(fundamentals_dir / "buybacks.csv", index=False)
            print(f"  Saved {len(shares_ttm)} quarters of shares")

    eps = eps_q
    if eps is not None:
        eps_ttm = _ttm_sum(eps).dropna()
        if not eps_ttm.empty:
            _quarter_frame(eps_ttm, "eps_diluted").to_csv(fundamentals_dir / "eps.csv", index=False)
            print(f"  Saved {len(eps_ttm)} quarters of EPS")

    fcf = fcf_q
    if fcf is not None:
        fcf_ttm = _ttm_sum(fcf).dropna()
        if not fcf_ttm.empty:
            _quarter_frame(fcf_ttm, "free_cash_flow").to_csv(fundamentals_dir / "fcf.csv", index=False)
            print(f"  Saved {len(fcf_ttm)} quarters of FCF")

    print(f"\nDone generating data for {ticker.upper()}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--ticker", "-t", required=True)
    parser.add_argument("--data-dir", default="data")
    parser.add_argument("--api-key", default=None)
    args = parser.parse_args()
    generate_data(args.ticker, Path(args.data_dir), api_key=args.api_key)

File: src/lib/generateData.js
import YahooFinance from 'yahoo-finance2';
const yahooFinance = new YahooFinance({ suppressNotices: ['ripHistorical'] });
import { supabase } from './supabase';

const ALPHA_VANTAGE_BASE_URL = 'https://www.alphavantage.co/query';

function quarterLabel(date) {
  const month = date.getMonth() + 1;
  const quarter = Math.ceil(month / 3);
  return `Q${quarter}`;
}

function quarterFrame(entries, valueName) {
  return entries.map(e => ({
    year: e.date.getFullYear(),
    quarter: quarterLabel(e.date),
    [valueName]: e.value,
  }));
}

function ttmSum(values) {
  const result = [];
  for (let i = 3; i < values.length; i++) {
    const sum = values[i].value + values[i - 1].value + values[i - 2].value + values[i - 3].value;
    result.push({ date: values[i].date, value: sum });
  }
  return result;
}

function ttmMean(values) {
  const result = [];
  for (let i = 3; i < values.length; i++) {
    const avg = (values[i].value + values[i - 1].value + values[i - 2].value + values[i - 3].value) / 4;
    result.push({ date: values[i].date, value: avg });
  }
  return result;
}

async function fetchAlphaVantage(func, symbol, apiKey) {
  const url = `${ALPHA_VANTAGE_BASE_URL}?function=${func}&symbol=${symbol}&apikey=${apiKey}`;
  const res = await fetch(url, { signal: AbortSignal.timeout(30000) });
  const data = await res.json();

  if (data['Error Message']) throw new Error(`Alpha Vantage API error: ${data['Error Message']}`);
  if (data['Note']) throw new Error(`Alpha Vantage rate limit: ${data['Note']}`);
  if (data['Information']) throw new Error(`Alpha Vantage: ${data['Information']}`);
  return data;
}

function parseReports(data, key) {
  const reports = data[key] || [];
  return reports
    .filter(r => r.fiscalDateEnding)
    .map(r => {
      const parsed = { date: new Date(r.fiscalDateEnding) };
      for (const [k, v] of Object.entries(r)) {
        if (k === 'fiscalDateEnding' || k === 'reportedCurrency') continue;
        const num = parseFloat(v);
        parsed[k] = isNaN(num) ? null : num;
      }
      return parsed;
    })
    .sort((a, b) => a.date - b.date);
}

function pickSeries(reports, fieldNames) {
  for (const name of fieldNames) {
    const values = reports
      .filter(r => r[name] != null && !isNaN(r[name]))
      .map(r => ({ date: r.date, value: r[name] }));
    if (values.length > 0) return values;
  }
  return null;
}

function mergeQA(quarterly, annual) {
  if (!quarterly && !annual) return null;
  if (!quarterly) return annual;
  if (!annual) return quarterly;
  const earliestQ = quarterly[0].date;
  const before = annual.filter(a => a.date < earliestQ);
  return [...before, ...quarterly].sort((a, b) => a.date - b.date);
}

async function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export async function generateTickerData(ticker, apiKey) {
  const upper = ticker.toUpperCase();

  // Fetch prices from yahoo-finance2
  const chartResult = await yahooFinance.chart(upper, { period1: '1990-01-01', period2: new Date().toISOString().split('T')[0] });
  const hist = chartResult.quotes || [];
  if (!hist || hist.length === 0) throw new Error(`No price data for ${upper}`);

  const dailyPrices = hist
    .filter(h => h.close != null)
    .map(h => ({
      date: h.date instanceof Date
        ? h.date.toISOString().split('T')[0]
        : new Date(h.date).toISOString().split('T')[0],
      close: h.close,
    }));

  // Market data metrics
  const currentPrice = dailyPrices[dailyPrices.length - 1].close;
  const currentDate = dailyPrices[dailyPrices.length - 1].date;
  const last252 = dailyPrices.slice(-252);
  const week52High = Math.max(...last252.map(p => p.close));
  const week52Low = Math.min(...last252.map(p => p.close));
  const pctFromHigh = ((currentPrice - week52High) / week52High) * 100;

  const marketData = [
    { metric: 'current_price', value: currentPrice, date: currentDate },
    { metric: '52_week_high', value: week52High, date: currentDate },
    { metric: '52_week_low', value: week52Low, date: currentDate },
    { metric: 'pct_from_52week_high', value: Math.round(pctFromHigh * 100) / 100, date: currentDate },
  ];

  if (dailyPrices.length >= 252) {
    const price1yAgo = dailyPrices[dailyPrices.length - 252].close;
    const pct1y = ((currentPrice - price1yAgo) / price1yAgo) * 100;
    marketData.push({ metric: 'pct_change_1y', value: Math.round(pct1y * 100) / 100, date: currentDate });
  }

  // Save prices to Supabase
  await supabase.from('ticker_prices').upsert([
    { ticker: upper, data_type: 'daily_prices', data: dailyPrices, updated_at: new Date().toISOString() },
    { ticker: upper, data_type: 'market_data', data: marketData, updated_at: new Date().toISOString() },
  ]);

  // Fetch fundamentals from Alpha Vantage
  const incomeRaw = await fetchAlphaVantage('INCOME_STATEMENT', upper, apiKey);
  const incomeQ = parseReports(incomeRaw, 'quarterlyReports');
  const incomeA = parseReports(incomeRaw, 'annualReports');
  await sleep(12000);

  const balanceRaw = await fetchAlphaVantage('BALANCE_SHEET', upper, apiKey);
  const balanceQ = parseReports(balanceRaw, 'quarterlyReports');
  const balanceA = parseReports(balanceRaw, 'annualReports');
  await sleep(12000);

  const cashRaw = await fetchAlphaVantage('CASH_FLOW', upper, apiKey);
  const cashQ = parseReports(cashRaw, 'quarterlyReports');
  const cashA = parseReports(cashRaw, 'annualReports');

  // Extract series
  let revenueQ = pickSeries(incomeQ, ['totalRevenue']);
  const revenueA = pickSeries(incomeA, ['totalRevenue']);

  const operatingIncomeQ = pickSeries(incomeQ, ['operatingIncome']);

  let epsQ = pickSeries(incomeQ, ['dilutedEPS', 'reportedEPS']);
  if (!epsQ) {
    const netIncome = pickSeries(incomeQ, ['netIncome', 'netIncomeFromContinuingOperations']);
    const dilutedShares = pickSeries(incomeQ, ['dilutedAverageShares', 'dilutedAverageSharesOutstanding']);
    const sharesOutQ = pickSeries(balanceQ, ['commonStockSharesOutstanding', 'commonStock']);
    if (netIncome && (dilutedShares || sharesOutQ)) {
      const denom = dilutedShares || sharesOutQ;
      epsQ = netIncome.map(n => {
        const d = denom.find(s => s.date.getTime() === n.date.getTime());
        return d ? { date: n.date, value: n.value / d.value } : null;
      }).filter(Boolean);
    }
  }

  let fcfQ = pickSeries(cashQ, ['freeCashFlow']);
  if (!fcfQ) {
    const opCf = pickSeries(cashQ, ['operatingCashflow']);
    const capex = pickSeries(cashQ, ['capitalExpenditures']);
    if (opCf && capex) {
      fcfQ = opCf.map(o => {
        const c = capex.find(x => x.date.getTime() === o.date.getTime());
        return c ? { date: o.date, value: o.value - c.value } : null;
      }).filter(Boolean);
    }
  }

  const sharesOutQ = pickSeries(balanceQ, ['commonStockSharesOutstanding', 'commonStock']);

  const fundamentalUpserts = [];

  // Revenue
  const revenue = revenueQ;
  if (revenue && revenue.length >= 4) {
    const ttm = ttmSum(revenue);
    if (ttm.length > 0) {
      fundamentalUpserts.push({
        ticker: upper,
        data_type: 'revenue',
        data: quarterFrame(ttm, 'revenue'),
        updated_at: new Date().toISOString(),
      });
    }
  }

  // Operating margins
  if (operatingIncomeQ && revenue && revenue.length >= 4 && operatingIncomeQ.length >= 4) {
    const ttmOp = ttmSum(operatingIncomeQ);
    const ttmRev = ttmSum(revenue);
    const margins = [];
    for (let i = 0; i < ttmOp.length; i++) {
      const revEntry = ttmRev.find(r => r.date.getTime() === ttmOp[i].date.getTime());
      if (revEntry && revEntry.value !== 0) {
        margins.push({ date: ttmOp[i].date, value: ttmOp[i].value / revEntry.value });
      }
    }
    if (margins.length > 0) {
      fundamentalUpserts.push({
        ticker: upper,
        data_type: 'operating_margins',
        data: quarterFrame(margins, 'operating_margin'),
        updated_at: new Date().toISOString(),
      });
    }
  }

  // Shares outstanding (buybacks)
  if (sharesOutQ && sharesOutQ.length >= 4) {
    const ttm = ttmMean(sharesOutQ);
    if (ttm.length > 0) {
      fundamentalUpserts.push({
        ticker: upper,
        data_type: 'buybacks',
        data: quarterFrame(ttm, 'shares_outstanding'),
        updated_at: new Date().toISOString(),
      });
    }
  }

  // EPS
  if (epsQ && epsQ.length >= 4) {
    const ttm = ttmSum(epsQ);
    if (ttm.length > 0) {
      fundamentalUpserts.push({
        ticker: upper,
        data_type: 'eps',
        data: quarterFrame(ttm, 'eps_diluted'),
        updated_at: new Date().toISOString(),
      });
    }
  }

  // FCF
  if (fcfQ && fcfQ.length >= 4) {
    const ttm = ttmSum(fcfQ);
    if (ttm.length > 0) {
      fundamentalUpserts.push({
        ticker: upper,
        data_type: 'fcf',
        data: quarterFrame(ttm, 'free_cash_flow'),
        updated_at: new Date().toISOString(),
      });
    }
  }

  if (fundamentalUpserts.length > 0) {
    const { error } = await supabase.from('ticker_fundamentals').upsert(fundamentalUpserts);
    if (error) throw new Error(`Failed to save fundamentals: ${error.message}`);
  }

  return { ticker: upper, pricesDays: dailyPrices.length, fundamentalsTypes: fundamentalUpserts.length };
}
